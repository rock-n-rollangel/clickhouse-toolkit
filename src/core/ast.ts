/**
 * Core AST (Abstract Syntax Tree) definitions
 * Represents the structure of queries independent of SQL text
 */

export type Primitive = string | number | boolean | Date | null

export interface ColumnRef {
  type: 'column'
  name: string
  table?: string
  alias?: string
}

export interface Value {
  type: 'value'
  value: Primitive
}

export interface ArrayValue {
  type: 'array'
  values: Primitive[]
}

export interface TupleValue {
  type: 'tuple'
  values: Primitive[]
}

export interface Subquery {
  type: 'subquery'
  query: SelectNode
}

export type Expr = ColumnRef | Value | ArrayValue | TupleValue | Subquery

export interface Predicate {
  type: 'predicate'
  left: Expr
  operator: string
  right: Expr
}

export interface AndPredicate {
  type: 'and'
  predicates: PredicateNode[]
  fromCombinator?: boolean
}

export interface OrPredicate {
  type: 'or'
  predicates: PredicateNode[]
}

export interface NotPredicate {
  type: 'not'
  predicate: PredicateNode
}

export type PredicateNode = Predicate | AndPredicate | OrPredicate | NotPredicate

export interface OrderSpec {
  column: ColumnRef
  direction: 'ASC' | 'DESC'
}

export interface JoinSpec {
  type: 'inner' | 'left' | 'right' | 'full'
  table: string | Subquery
  alias?: string
  on: PredicateNode
}

export interface WithClause {
  alias: string
  query: SelectNode
}

export interface SelectNode {
  type: 'select'
  columns: ColumnRef[]
  from?: {
    table: string | Subquery
    alias?: string
  }
  joins?: JoinSpec[]
  with?: WithClause[]
  prewhere?: PredicateNode
  where?: PredicateNode
  groupBy?: ColumnRef[]
  having?: PredicateNode
  orderBy?: OrderSpec[]
  limit?: number
  offset?: number
  final?: boolean
  settings?: Record<string, any>
  format?: string
}

export interface InsertNode {
  type: 'insert'
  table: string
  columns: string[]
  values: Primitive[][]
  format?: string
}

export interface UpdateNode {
  type: 'update'
  table: string
  set: Record<string, Primitive>
  where?: PredicateNode
  settings?: Record<string, any>
}

export interface DeleteNode {
  type: 'delete'
  table: string
  where?: PredicateNode
  settings?: Record<string, any>
}

export type QueryNode = SelectNode | InsertNode | UpdateNode | DeleteNode
