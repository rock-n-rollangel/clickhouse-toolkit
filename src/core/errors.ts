/**
 * Typed error system for ClickHouse Toolkit
 * Machine-readable error codes with context
 */

export abstract class ClickHouseToolkitError extends Error {
  abstract readonly code: string
  abstract readonly type: string
  readonly queryId?: string
  readonly timestamp: Date

  constructor(message: string, queryId?: string) {
    super(message)
    this.name = this.constructor.name
    this.timestamp = new Date()
    this.queryId = queryId
  }

  toJSON() {
    return {
      name: this.name,
      code: this.code,
      type: this.type,
      message: this.message,
      queryId: this.queryId,
      timestamp: this.timestamp.toISOString(),
    }
  }
}

// Query execution errors
export class QueryError extends ClickHouseToolkitError {
  readonly code = 'QUERY_ERROR'
  readonly type = 'execution'
  readonly clickhouseCode?: string
  readonly sql?: string

  constructor(message: string, queryId?: string, clickhouseCode?: string, sql?: string) {
    super(message, queryId)
    this.clickhouseCode = clickhouseCode
    this.sql = sql
  }
}

export class TimeoutError extends ClickHouseToolkitError {
  readonly code = 'TIMEOUT_ERROR'
  readonly type = 'execution'
  readonly timeoutMs?: number

  constructor(message: string, queryId?: string, timeoutMs?: number) {
    super(message, queryId)
    this.timeoutMs = timeoutMs
  }
}

export class CancelledError extends ClickHouseToolkitError {
  readonly code = 'CANCELLED_ERROR'
  readonly type = 'execution'

  constructor(message: string, queryId?: string) {
    super(message, queryId)
  }
}

// Validation errors
export class ValidationError extends ClickHouseToolkitError {
  readonly code = 'VALIDATION_ERROR'
  readonly type = 'validation'
  readonly field?: string
  readonly value?: any

  constructor(message: string, queryId?: string, field?: string, value?: any) {
    super(message, queryId)
    this.field = field
    this.value = value
  }
}

export class IdentifierError extends ClickHouseToolkitError {
  readonly code = 'IDENTIFIER_ERROR'
  readonly type = 'validation'
  readonly identifier: string

  constructor(message: string, identifier: string, queryId?: string) {
    super(message, queryId)
    this.identifier = identifier
  }
}

export class ParameterError extends ClickHouseToolkitError {
  readonly code = 'PARAMETER_ERROR'
  readonly type = 'validation'
  readonly parameter: string

  constructor(message: string, parameter: string, queryId?: string) {
    super(message, queryId)
    this.parameter = parameter
  }
}

// Migration errors
export class MigrationError extends ClickHouseToolkitError {
  readonly code = 'MIGRATION_ERROR'
  readonly type = 'migration'
  readonly migrationId?: string
  readonly step?: string

  constructor(message: string, queryId?: string, migrationId?: string, step?: string) {
    super(message, queryId)
    this.migrationId = migrationId
    this.step = step
  }
}

export class DriftError extends ClickHouseToolkitError {
  readonly code = 'DRIFT_ERROR'
  readonly type = 'migration'
  readonly differences: Array<{
    type: string
    table: string
    column?: string
    expected?: any
    actual?: any
  }>

  constructor(message: string, differences: any[], queryId?: string) {
    super(message, queryId)
    this.differences = differences
  }
}

// Connection errors
export class ConnectionError extends ClickHouseToolkitError {
  readonly code = 'CONNECTION_ERROR'
  readonly type = 'connection'
  readonly url?: string

  constructor(message: string, queryId?: string, url?: string) {
    super(message, queryId)
    this.url = url
  }
}

// Schema errors
export class SchemaError extends ClickHouseToolkitError {
  readonly code = 'SCHEMA_ERROR'
  readonly type = 'schema'
  readonly table?: string
  readonly column?: string

  constructor(message: string, queryId?: string, table?: string, column?: string) {
    super(message, queryId)
    this.table = table
    this.column = column
  }
}

// Error factory functions
export function createQueryError(message: string, queryId?: string, clickhouseCode?: string, sql?: string): QueryError {
  return new QueryError(message, queryId, clickhouseCode, sql)
}

export function createTimeoutError(message: string, queryId?: string, timeoutMs?: number): TimeoutError {
  return new TimeoutError(message, queryId, timeoutMs)
}

export function createValidationError(message: string, queryId?: string, field?: string, value?: any): ValidationError {
  return new ValidationError(message, queryId, field, value)
}

export function createIdentifierError(message: string, identifier: string, queryId?: string): IdentifierError {
  return new IdentifierError(message, identifier, queryId)
}

export function createParameterError(message: string, parameter: string, queryId?: string): ParameterError {
  return new ParameterError(message, parameter, queryId)
}

export function createMigrationError(
  message: string,
  queryId?: string,
  migrationId?: string,
  step?: string,
): MigrationError {
  return new MigrationError(message, queryId, migrationId, step)
}

export function createConnectionError(message: string, queryId?: string, url?: string): ConnectionError {
  return new ConnectionError(message, queryId, url)
}

export function createSchemaError(message: string, queryId?: string, table?: string, column?: string): SchemaError {
  return new SchemaError(message, queryId, table, column)
}
