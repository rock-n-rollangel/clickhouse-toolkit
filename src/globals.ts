import { MetadataArgsStorage } from './metadata/metadata-args-storage'

export function getMetadataArgsStorage(): MetadataArgsStorage {
  if (!global.metadataStorage) global.metadataStorage = new MetadataArgsStorage()
  return global.metadataStorage
}
