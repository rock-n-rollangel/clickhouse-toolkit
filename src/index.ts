export * from './connection/connection'
export * from './decorators/schema/schema'
export * from './decorators/schema/options/schema-options-base'
export * from './decorators/schema/options/schema-options'
export * from './decorators/schema/options/materialized-options'
export * from './decorators/column/column'
export * from './common/column-options'
export * from './query-runner/query-runner'
export * from './query-builder/delete-query-builder'
export * from './query-builder/select-query-builder'
export * from './query-builder/insert-query-builder'
export * from './query-builder/update-query-builder'
export * from './query-builder/query-builder'
export * from './metadata/schema-metadata'
export * from './types/engine'
export * from './types/data-types'
export * from './types/base-table-options'
export * from './types/column-schema'
export * from './types/connection-options'
export * from './types/materialized-view-options'
export * from './types/object-literal'
export * from './types/params'
export * from './types/table-options'
export * from './types/table-schema'
export * from './types/where-clause-type'
export * from './types/where-clause'
export * from './schema-builder/table'
