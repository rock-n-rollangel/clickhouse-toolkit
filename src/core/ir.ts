/**
 * IR (Intermediate Representation) definitions
 * Normalized, minimal data structure for rendering
 */

import { Primitive } from './ast'

// Forward declarations to break circular dependencies
export interface QueryIR {
  type: 'select' | 'insert' | 'update' | 'delete'
  table: string
  tableAlias?: string
  columns?: ExprIR[]
  predicates: NormalizedPredicateNode[]
  orderBy?: Array<{ column: string; direction: 'ASC' | 'DESC' }>
  limit?: number
  offset?: number
  final?: boolean
  settings?: Record<string, any>
  joins?: Array<{
    type: 'inner' | 'left' | 'right' | 'full'
    table: string
    alias?: string
    on: NormalizedPredicateNode
  }>
  with?: Array<{
    alias: string
    query: QueryIR
  }>
  groupBy?: string[]
  having?: NormalizedPredicateNode[]
  values?: Array<Primitive[]>
  set?: Record<string, Primitive>
}

export interface NormalizedPredicate {
  type: 'predicate'
  left: string // normalized column reference
  operator: string
  right: Primitive | Primitive[]
  isPrewhere?: boolean
}

export interface NormalizedAndPredicate {
  type: 'and'
  predicates: NormalizedPredicateNode[]
  isPrewhere?: boolean
  fromCombinator?: boolean
}

export interface NormalizedOrPredicate {
  type: 'or'
  predicates: NormalizedPredicateNode[]
  isPrewhere?: boolean
}

export interface NormalizedNotPredicate {
  type: 'not'
  predicate: NormalizedPredicateNode
  isPrewhere?: boolean
}

export interface RawPredicateIR {
  type: 'raw_predicate'
  sql: string
}

export type NormalizedPredicateNode =
  | NormalizedPredicate
  | NormalizedAndPredicate
  | NormalizedOrPredicate
  | NormalizedNotPredicate
  | RawPredicateIR

interface BaseExprIR {
  alias?: string
}

interface ColumnExprIR extends BaseExprIR {
  exprType: 'column'
  columnName: string
  tableName?: string
}

interface RawExprIR extends BaseExprIR {
  exprType: 'raw'
  rawSql: string
}

interface FunctionExprIR extends BaseExprIR {
  exprType: 'function'
  functionName: string
  functionArgs: ExprIR[]
}

interface CaseExprIR extends BaseExprIR {
  exprType: 'case'
  caseCases: Array<{ condition: NormalizedPredicateNode; then: ExprIR }>
  caseElse?: ExprIR
}

interface ValueExprIR extends BaseExprIR {
  exprType: 'value'
  value: Primitive
}

interface SubqueryExprIR extends BaseExprIR {
  exprType: 'subquery'
  subquery: QueryIR
}

interface ArrayExprIR extends BaseExprIR {
  exprType: 'array'
  values: Primitive[]
}

interface TupleExprIR extends BaseExprIR {
  exprType: 'tuple'
  values: Primitive[]
}

export type ExprIR =
  | ColumnExprIR
  | RawExprIR
  | FunctionExprIR
  | CaseExprIR
  | ValueExprIR
  | SubqueryExprIR
  | ArrayExprIR
  | TupleExprIR

export interface NormalizedQuery {
  type: 'select' | 'insert' | 'update' | 'delete'
  table: string
  columns?: string[]
  predicates: NormalizedPredicate[]
  orderBy?: Array<{ column: string; direction: 'ASC' | 'DESC' }>
  limit?: number
  offset?: number
  final?: boolean
  settings?: Record<string, any>
  joins?: Array<{
    type: 'inner' | 'left' | 'right' | 'full'
    table: string
    alias?: string
    on: NormalizedPredicate
  }>
}

export interface QueryContext {
  queryId: string
  timestamp: Date
  settings: Record<string, any>
}

export interface ValidationResult {
  valid: boolean
  errors: string[]
  warnings: string[]
}
