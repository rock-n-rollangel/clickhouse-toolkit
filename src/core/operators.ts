/**
 * Operator helpers for building predicates
 * Values-first approach for safety
 */

import { Primitive } from './ast'

export interface Operator {
  type: string
  value: Primitive | Primitive[] | any // 'any' to support SelectBuilder without circular dependency
}

// Comparison operators
export function Eq(value: Primitive | any): Operator {
  return { type: 'eq', value }
}

export function EqCol(column: string): Operator {
  return { type: 'eq_col', value: column }
}

export function Ne(value: Primitive | any): Operator {
  return { type: 'ne', value }
}

export function Gt(value: Primitive | any): Operator {
  return { type: 'gt', value }
}

export function Gte(value: Primitive | any): Operator {
  return { type: 'gte', value }
}

export function Lt(value: Primitive | any): Operator {
  return { type: 'lt', value }
}

export function Lte(value: Primitive | any): Operator {
  return { type: 'lte', value }
}

// Set operators
export function In(values: Primitive[] | any): Operator {
  return { type: 'in', value: values }
}

export function NotIn(values: Primitive[]): Operator {
  return { type: 'not_in', value: values }
}

export function Between(start: Primitive, end: Primitive): Operator {
  return { type: 'between', value: [start, end] }
}

// String operators
export function Like(pattern: string): Operator {
  return { type: 'like', value: pattern }
}

export function ILike(pattern: string): Operator {
  return { type: 'ilike', value: pattern }
}

export function RegExp(pattern: string, flags?: string): Operator {
  return { type: 'regexp', value: { pattern, flags } as any }
}

// Null operators
export function IsNull(): Operator {
  return { type: 'is_null', value: null }
}

export function IsNotNull(): Operator {
  return { type: 'is_not_null', value: null }
}

// Array operators
export function HasAny(values: Primitive[]): Operator {
  return { type: 'has_any', value: values }
}

export function HasAll(values: Primitive[]): Operator {
  return { type: 'has_all', value: values }
}

// Tuple operators
export function InTuple(tupleRows: Primitive[][]): Operator {
  return { type: 'in_tuple', value: tupleRows as any }
}

// Subquery operators
export function Exists(subquery: any): Operator {
  return { type: 'exists', value: subquery }
}

export function NotExists(subquery: any): Operator {
  return { type: 'not_exists', value: subquery }
}

// Boolean combinators
export interface PredicateCombinator {
  type: 'and' | 'or' | 'not'
  predicates: (WhereInput | Operator)[]
}

export function And(...predicates: (WhereInput | Operator)[]): PredicateCombinator {
  return { type: 'and', predicates }
}

export function Or(...predicates: (WhereInput | Operator)[]): PredicateCombinator {
  return { type: 'or', predicates }
}

export function Not(predicate: WhereInput | Operator): PredicateCombinator {
  return { type: 'not', predicates: [predicate] }
}

// Escape hatch for raw SQL (explicitly unsafe)
export function Raw(sql: string): Operator {
  return { type: 'raw', value: sql }
}

// Helper functions for common patterns
export function startsWith(prefix: string): Operator {
  return Like(`${escapeLikePattern(prefix)}%`)
}

export function endsWith(suffix: string): Operator {
  return Like(`%${escapeLikePattern(suffix)}`)
}

export function contains(substring: string): Operator {
  return Like(`%${escapeLikePattern(substring)}%`)
}

function escapeLikePattern(pattern: string): string {
  return pattern.replace(/[%_\\]/g, '\\$&')
}

// Type definitions
export type WhereInput = Record<string, Operator | PredicateCombinator> | PredicateCombinator
