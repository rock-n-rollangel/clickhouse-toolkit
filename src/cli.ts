#!/usr/bin/env node
import 'ts-node/register'
import { ConnectionOptions } from './types/connection-options'
import { Connection } from './connection/connection'
import { ClickHouseToolkitError } from './errors/clickhouse-toolkit-error'

async function main() {
  const command = process.argv[2]
  const args = process.argv.slice(3)

  const options = {
    database: process.env.CLICKHOUSE_DATABASE,
    migrations: process.env.CLICKHOUSE_MIGRATIONS?.split(','),
    migrationsTableName: process.env.CLICKHOUSE_MIGRATIONS_TABLE_NAME,
    url: process.env.CLICKHOUSE_URL,
    username: process.env.CLICKHOUSE_USERNAME,
    password: process.env.CLICKHOUSE_PASSWORD,
    logging: true,
  } as ConnectionOptions

  if (!options.database || !options.migrations || !options.url || !options.username || !options.password) {
    throw new ClickHouseToolkitError('Missing required environment variables')
  }

  const connection = await Connection.initialize(options)

  await connection.migrator.init()

  switch (command) {
    case 'up':
      await connection.migrator.up()
      break
    case 'down':
      await connection.migrator.down(parseInt(args[0]))
      break
    case 'status':
      await connection.migrator.status()
      break
  }

  process.exit(0)
}

main()
