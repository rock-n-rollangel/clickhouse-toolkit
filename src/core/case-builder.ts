/**
 * CASE WHEN THEN builder helper
 */

import { WhereInput, Operator } from './operators'
import { PredicateNode, Expr, CaseExpr } from './ast'
import { operatorToPredicate, parseColumnRef } from './predicate-builder'
import { createValidationError } from './errors'

export class CaseBuilder {
  private cases: Array<{ condition: PredicateNode; then: Expr }> = []
  private maxDepth: number = 0

  when(condition: WhereInput | PredicateNode, thenExpr: Expr): this {
    // Check for nested case expressions and validate depth
    if (thenExpr.type === 'case') {
      this.maxDepth++
      if (this.maxDepth > 10) {
        throw createValidationError('Maximum CASE nesting depth (10) exceeded', undefined, 'case', 'nesting')
      }
    }

    // Convert condition to PredicateNode
    let predicateNode: PredicateNode
    if (typeof condition === 'object' && 'type' in condition) {
      predicateNode = condition as PredicateNode
    } else {
      predicateNode = this.convertWhereInputToPredicate(condition as WhereInput)
    }

    this.cases.push({ condition: predicateNode, then: thenExpr })
    return this
  }

  else(elseExpr: Expr): CaseExpr {
    return {
      type: 'case',
      cases: this.cases,
      else: elseExpr,
    }
  }

  private convertWhereInputToPredicate(input: WhereInput): PredicateNode {
    const entries = Object.entries(input)
    if (entries.length === 1) {
      const [column, operator] = entries[0]
      return operatorToPredicate(column, operator as Operator, parseColumnRef)
    }
    return {
      type: 'and',
      predicates: entries.map(([col, op]) => operatorToPredicate(col, op as Operator, parseColumnRef)),
    }
  }
}

export function Case(): CaseBuilder {
  return new CaseBuilder()
}

export { CaseExpr }
