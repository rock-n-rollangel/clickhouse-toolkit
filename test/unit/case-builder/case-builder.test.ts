/**
 * Case Builder Tests
 * Comprehensive tests for the Case() when/then/else functionality
 */

import { describe, it, expect } from '@jest/globals'
import { Case } from '../../../src/core/case-builder'
import { CaseExpr } from '../../../src/core/ast'
import { Eq, Gt, Lt, In, IsNull, IsNotNull, Like, EqCol, Lte } from '../../../src/core/operators'
import { Count, Sum, Concat, Upper, Raw, Column, Value } from '../../../src/index'

describe('CaseBuilder', () => {
  describe('Basic Functionality Tests', () => {
    it('should create simple case with string condition and literal values', () => {
      const result = Case()
        .when({ status: Eq('active') }, Value('Active User'))
        .else(Value('Inactive User'))

      expect(result.type).toBe('case')
      expect(result.cases).toHaveLength(1)
      expect(result.cases[0].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'status' },
        operator: '=',
        right: { type: 'value', value: 'active' },
      })
      expect(result.cases[0].then).toEqual(Value('Active User'))
      expect(result.else).toEqual(Value('Inactive User'))
    })

    it('should create case with literal values in then/else', () => {
      const result = Case()
        .when({ age: Gt(18) }, Value(1))
        .else(Value(0))

      expect(result.type).toBe('case')
      expect(result.cases[0].then).toEqual(Value(1))
      expect(result.else).toEqual(Value(0))
    })

    it('should create case with boolean values', () => {
      const result = Case()
        .when({ active: Eq(true) }, Value(true))
        .else(Value(false))

      expect(result.cases[0].then).toEqual(Value(true))
      expect(result.else).toEqual(Value(false))
    })

    it('should create case with null values', () => {
      const result = Case().when({ status: IsNull() }, Value(null)).else(Value('Has Status'))

      expect(result.cases[0].then).toEqual(Value(null))
      expect(result.else).toEqual(Value('Has Status'))
    })

    it('should handle multiple when clauses', () => {
      const result = Case()
        .when({ age: Lt(18) }, Value('Minor'))
        .when({ age: Gt(65) }, Value('Senior'))
        .else(Value('Adult'))

      expect(result.cases).toHaveLength(2)
      expect(result.cases[0].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'age' },
        operator: '<',
        right: { type: 'value', value: 18 },
      })
      expect(result.cases[1].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'age' },
        operator: '>',
        right: { type: 'value', value: 65 },
      })
      expect(result.cases[0].then).toEqual(Value('Minor'))
      expect(result.cases[1].then).toEqual(Value('Senior'))
      expect(result.else).toEqual(Value('Adult'))
    })

    it('should handle case with column references in then/else', () => {
      const result = Case()
        .when({ priority: Eq('high') }, Column('priority_level'))
        .else(Column('normal_level'))

      expect(result.cases[0].then).toEqual(Column('priority_level'))
      expect(result.else).toEqual(Column('normal_level'))
    })

    it('should properly distinguish between Column() and Value() usage', () => {
      const columnCase = Case()
        .when({ priority: Eq('high') }, Column('priority_level'))
        .else(Column('normal_level'))

      const valueCase = Case()
        .when({ priority: Eq('high') }, Value('Priority: High'))
        .else(Value('Priority: Normal'))

      expect(columnCase.cases[0].then).toEqual(Column('priority_level'))
      expect(columnCase.else).toEqual(Column('normal_level'))
      expect(valueCase.cases[0].then).toEqual(Value('Priority: High'))
      expect(valueCase.else).toEqual(Value('Priority: Normal'))
    })

    it('should demonstrate proper Value() usage for different literal types', () => {
      const result = Case()
        .when({ status: Eq('active') }, Value('Active User'))
        .when({ age: Gt(65) }, Value('Senior Citizen'))
        .when({ verified: Eq(true) }, Value(1))
        .when({ disabled: Eq(false) }, Value(0))
        .else(Value(null))

      expect(result.cases[0].then).toEqual(Value('Active User'))
      expect(result.cases[1].then).toEqual(Value('Senior Citizen'))
      expect(result.cases[2].then).toEqual(Value(1))
      expect(result.cases[3].then).toEqual(Value(0))
      expect(result.else).toEqual(Value(null))
    })
  })

  describe('Advanced Features Tests', () => {
    it('should handle nested case expressions', () => {
      const innerCase = Case()
        .when({ type: Eq('premium') }, Value('Premium Member'))
        .else(Value('Regular Member'))

      const result = Case()
        .when({ active: Eq(true) }, innerCase)
        .else(Value('Inactive'))

      expect(result.type).toBe('case')
      expect(result.cases[0].then.type).toBe('case')
      expect((result.cases[0].then as CaseExpr).cases).toHaveLength(1)
    })

    it('should handle case with SQL functions in then/else', () => {
      const result = Case()
        .when({ status: Eq('active') }, Count())
        .else(Sum('amount'))

      expect(result.cases[0].then).toEqual({
        type: 'function',
        name: 'count',
        args: [{ type: 'raw', sql: '*' }],
      })
      expect(result.else).toEqual({
        type: 'function',
        name: 'sum',
        args: [{ type: 'column', name: 'amount' }],
      })
    })

    it('should handle case with string functions', () => {
      const result = Case()
        .when({ name: Like('John%') }, Upper('first_name'))
        .else(Concat('Unknown', 'User'))

      expect(result.cases[0].then).toEqual({
        type: 'function',
        name: 'upper',
        args: [{ type: 'column', name: 'first_name' }],
      })
      expect(result.else).toEqual({
        type: 'function',
        name: 'concat',
        args: [
          { type: 'column', name: 'Unknown' },
          { type: 'column', name: 'User' },
        ],
      })
    })

    it('should handle complex predicates with And in conditions', () => {
      const result = Case()
        .when(
          {
            age: Gt(18),
            status: Eq('active'),
          },
          Value('Eligible'),
        )
        .else(Value('Not Eligible'))

      expect(result.cases[0].condition).toMatchObject({
        type: 'and',
        predicates: [
          {
            type: 'predicate',
            left: { type: 'column', name: 'age' },
            operator: '>',
            right: { type: 'value', value: 18 },
          },
          {
            type: 'predicate',
            left: { type: 'column', name: 'status' },
            operator: '=',
            right: { type: 'value', value: 'active' },
          },
        ],
      })
    })

    it('should handle Raw expressions in case statements', () => {
      const result = Case()
        .when({ status: Eq('active') }, Raw('NOW()'))
        .else(Raw('NULL'))

      expect(result.cases[0].then).toEqual(Raw('NOW()'))
      expect(result.else).toEqual(Raw('NULL'))
    })

    it('should handle column-to-column comparisons with EqCol', () => {
      const result = Case()
        .when({ 'a.value': EqCol('b.value') }, Value('Match'))
        .else(Value('No Match'))

      expect(result.cases[0].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'value', table: 'a' },
        operator: '=',
        right: { type: 'column', name: 'value', table: 'b' },
      })
    })

    it('should handle In operator conditions', () => {
      const result = Case()
        .when({ status: In(['active', 'pending']) }, Value('In Progress'))
        .else(Value('Unknown'))

      expect(result.cases[0].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'status' },
        operator: 'IN',
        right: {
          type: 'array',
          values: ['active', 'pending'],
        },
      })
    })

    it('should handle IsNull and IsNotNull operators', () => {
      const result = Case()
        .when({ email: IsNull() }, Value('No Email'))
        .when({ email: IsNotNull() }, Value('Has Email'))
        .else(Value('Unknown'))

      expect(result.cases).toHaveLength(2)
      expect(result.cases[0].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'email' },
        operator: 'IS NULL',
      })
      expect(result.cases[1].condition).toMatchObject({
        type: 'predicate',
        left: { type: 'column', name: 'email' },
        operator: 'IS NOT NULL',
      })
    })
  })

  describe('Edge Cases & Error Handling', () => {
    it('should handle case without else clause', () => {
      const result = Case()
        .when({ status: Eq('active') }, Value('Active'))
        .else(Value(null))

      expect(result.else).toEqual(Value(null))
    })

    it('should handle undefined else value', () => {
      const result = Case()
        .when({ status: Eq('active') }, Value('Active'))
        .else(Value(undefined))

      expect(result.else).toEqual(Value(undefined))
    })

    it('should throw error for maximum nesting depth exceeded', () => {
      let nestedCase = Case()
        .when({ a: Eq(1) }, Value('a'))
        .else(Value('not_a'))

      for (let i = 0; i < 10; i++) {
        const previousCase = nestedCase
        nestedCase = Case()
          .when({ [`level_${i}`]: Eq(1) }, previousCase)
          .else(Value(`fallback_${i}`))
      }

      expect(nestedCase.type).toBe('case')
    })

    it('should handle empty string values', () => {
      const result = Case()
        .when({ status: Eq('') }, Value('Empty Status'))
        .else(Value('Has Status'))

      expect(result.cases[0].condition).toMatchObject({
        right: { type: 'value', value: '' },
      })
      expect(result.cases[0].then).toEqual(Value('Empty Status'))
    })

    it('should handle zero and false values', () => {
      const result = Case()
        .when({ count: Eq(0) }, Value(false))
        .when({ active: Eq(false) }, Value(0))
        .else(Value(1))

      expect(result.cases).toHaveLength(2)
      expect(result.cases[0].condition).toMatchObject({
        right: { type: 'value', value: 0 },
      })
      expect(result.cases[1].condition).toMatchObject({
        right: { type: 'value', value: false },
      })
      expect(result.cases[0].then).toEqual(Value(false))
      expect(result.cases[1].then).toEqual(Value(0))
    })

    it('should handle date values', () => {
      const testDate = new Date('2023-01-01')
      const result = Case()
        .when({ created_at: Gt(testDate) }, Value('Recent'))
        .else(Value('Old'))

      expect(result.cases[0].condition).toMatchObject({
        right: { type: 'value', value: testDate },
      })
    })

    it('should handle complex nested case with multiple conditions', () => {
      const innerCase = Case()
        .when({ type: Eq('premium') }, Value('Premium User'))
        .when({ type: Eq('basic') }, Value('Basic User'))
        .else(Value('Unknown Type'))

      const result = Case()
        .when({ active: Eq(true) }, innerCase)
        .when({ suspended: Eq(true) }, Value('Suspended'))
        .else(Value('Inactive'))

      expect(result.cases).toHaveLength(2)
      expect((result.cases[0].then as CaseExpr).type).toBe('case')
      expect((result.cases[0].then as CaseExpr).cases).toHaveLength(2)
    })

    it('should handle predicate nodes directly as conditions', () => {
      const predicateNode = {
        type: 'predicate' as const,
        left: { type: 'column' as const, name: 'score' },
        operator: '>',
        right: { type: 'value' as const, value: 80 },
      }

      const result = Case().when(predicateNode, Value('High Score')).else(Value('Low Score'))

      expect(result.cases[0].condition).toBe(predicateNode)
    })
  })

  describe('Real-world Examples', () => {
    it('should handle user categorization by age ranges', () => {
      const result = Case()
        .when({ age: Lt(13) }, Value('Child'))
        .when({ age: Lt(18) }, Value('Teenager'))
        .when({ age: Lte(65) }, Value('Adult'))
        .else(Value('Senior'))

      expect(result.cases).toHaveLength(3)
    })

    it('should handle status mapping (numeric to string)', () => {
      const result = Case()
        .when({ status_code: Eq(1) }, Value('Active'))
        .when({ status_code: Eq(2) }, Value('Pending'))
        .when({ status_code: Eq(3) }, Value('Completed'))
        .when({ status_code: Eq(0) }, Value('Inactive'))
        .else(Value('Unknown Status'))

      expect(result.cases).toHaveLength(4)
      expect(result.cases.map((c) => c.condition)).toMatchSnapshot()
    })

    it('should handle conditional aggregation', () => {
      const result = Case()
        .when({ category: Eq('premium') }, Sum('revenue'))
        .when({ category: Eq('basic') }, Count())
        .else(Raw('0'))

      expect(result.cases[0].then).toMatchObject({
        type: 'function',
        name: 'sum',
        args: [{ type: 'column', name: 'revenue' }],
      })
      expect(result.cases[1].then).toMatchObject({
        type: 'function',
        name: 'count',
      })
      expect(result.else).toEqual(Raw('0'))
    })
  })
})
