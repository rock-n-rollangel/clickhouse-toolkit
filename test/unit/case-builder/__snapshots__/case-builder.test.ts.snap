// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseBuilder Real-world Examples should handle status mapping (numeric to string) 1`] = `
[
  {
    "left": {
      "name": "status_code",
      "type": "column",
    },
    "operator": "=",
    "right": {
      "type": "value",
      "value": 1,
    },
    "type": "predicate",
  },
  {
    "left": {
      "name": "status_code",
      "type": "column",
    },
    "operator": "=",
    "right": {
      "type": "value",
      "value": 2,
    },
    "type": "predicate",
  },
  {
    "left": {
      "name": "status_code",
      "type": "column",
    },
    "operator": "=",
    "right": {
      "type": "value",
      "value": 3,
    },
    "type": "predicate",
  },
  {
    "left": {
      "name": "status_code",
      "type": "column",
    },
    "operator": "=",
    "right": {
      "type": "value",
      "value": 0,
    },
    "type": "predicate",
  },
]
`;
